crates/shielded_token/src/validation.rs

```diff
@@ -0,0 +1,350 @@
+pub const CONVERT_NAME: &str = "masp-convert.params";
+
+/// Env var to point to a dir with MASP parameters. When not specified,
+/// the default OS specific path is used.
+pub const ENV_VAR_MASP_PARAMS_DIR: &str = "NAMADA_MASP_PARAMS_DIR";
+
+/// Get the path to MASP parameters from [`ENV_VAR_MASP_PARAMS_DIR`] env var or
+/// use the default.
+pub fn get_params_dir() -> PathBuf {
+    if let Ok(params_dir) = env::var(ENV_VAR_MASP_PARAMS_DIR) {
+        #[allow(clippy::print_stdout)]
+        {
+            println!("Using {} as masp parameter folder.", params_dir);
+        }
+        PathBuf::from(params_dir)
+    } else {
+        masp_proofs::default_params_folder().unwrap()
+    }
+}
+
+/// Represents an authorization where the Sapling bundle is authorized and the
+/// transparent bundle is unauthorized.
+pub struct PartialAuthorized;
+
+impl Authorization for PartialAuthorized {
+    type SaplingAuth = <Authorized as Authorization>::SaplingAuth;
+    type TransparentAuth = <Unauthorized as Authorization>::TransparentAuth;
+}
+
+/// MASP verifying keys
+pub struct PVKs {
+    /// spend verifying key
+    pub spend_vk: VerifyingKey<Bls12>,
+    /// convert verifying key
+    pub convert_vk: VerifyingKey<Bls12>,
+    /// output verifying key
+    pub output_vk: VerifyingKey<Bls12>,
+}
+
+lazy_static! {
+    /// MASP verifying keys load from parameters
+    static ref VERIFIYING_KEYS: PVKs =
+        {
+        let params_dir = get_params_dir();
+        let [spend_path, convert_path, output_path] =
+            [SPEND_NAME, CONVERT_NAME, OUTPUT_NAME].map(|p| params_dir.join(p));
+
+        if !spend_path.exists() || !convert_path.exists() || !output_path.exists() {
+            let paths = masp_proofs::download_masp_parameters(None).expect(
+                "MASP parameters were not present, expected the download to \
+                succeed",
+            );
+            if paths.spend != spend_path
+                || paths.convert != convert_path
+                || paths.output != output_path
+            {
+                panic!(
+                    "unrecoverable: downloaded missing masp params, but to an \
+                    unfamiliar path"
+                )
+            }
+        }
+        // size and blake2b checked here
+        let params = masp_proofs::load_parameters(
+            spend_path.as_path(),
+            output_path.as_path(),
+            convert_path.as_path(),
+        );
+        PVKs {
+            spend_vk: params.spend_params.vk,
+            convert_vk: params.convert_params.vk,
+            output_vk: params.output_params.vk
+        }
+    };
+}
+
+/// Make sure the MASP params are present and load verifying keys into memory
+pub fn preload_verifying_keys() -> &'static PVKs {
+    &VERIFIYING_KEYS
+}
+
+fn load_pvks() -> &'static PVKs {
+    &VERIFIYING_KEYS
+}
+
+/// Verify a shielded transaction.
+pub fn verify_shielded_tx<F>(
+    transaction: &Transaction,
+    consume_verify_gas: F,
```

Comment:
Do we have access to the `VpGasMeter` object from this crate. Would be a bit cleaner if we could pass a mutable reference directly to it, but no worries if we don't have it

Reply:
yeah we do and I tried to use it, but because the `VpEnv::charge_gas` replaces the error with `meter.consume(used_gas).map_err(|_| { Error::SimpleMessage("Gas limit exceeded in native vp")})` it actually makes it more verbose :/ I'd like to change how to handle composition of error types with #1214 that could make this nicer, but for now I'd leave it as is